Source/WebCore/dom/MessagePort.cpp

```diff
@@ -350,14 +350,17 @@ bool MessagePort::virtualHasPendingActivity() const
     if (!context || m_isDetached)
         return false;
 
-    // If this object has been idle since the remote port declared itself elgibile for GC, we can GC.
-    if (!m_hasHadLocalActivitySinceLastCheck && m_isRemoteEligibleForGC)
-        return false;
-
     // If this MessagePort has no message event handler then the existence of remote activity cannot keep it alive.
     if (!m_hasMessageEventListener)
         return false;
 
+    if (m_entangled)
+        return true;
+
+    // If this object has been idle since the remote port declared itself elgibile for GC, we can GC.
+    if (!m_hasHadLocalActivitySinceLastCheck && m_isRemoteEligibleForGC)
+        return false;
```

Comment:
I'm pretty sure that the logic of m_hasHadLocalActivitySinceLastCheck and m_mightBeEligibleForGC is incorrect. It's not mentioned in the spec. And it doesn't make intuitive sense.

m_isRemoteEligibleForGC, isEntangled(), and m_hasMessageEventListener should cover everything. Maybe m_hasHadLocalActivitySinceLastCheck and m_mightBeEligibleForGC can be removed.

Reply:
Side note: I would rename m_isRemoteEligibleForGC to m_remoteHasActivity.

Reply:
I agree the current code is very complicated and likely not entirely correct. I was planning on following up to attempt to clean this up since this is a little riskier and might be a bigger change too).